{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2310ae4a-a06b-4d8a-8137-955f6e93139c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "الشكل العام للبيانات:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>...</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>SomaticClinicalImpact</th>\n",
       "      <th>SomaticClinicalImpactLastEvaluated</th>\n",
       "      <th>ReviewStatusClinicalImpact</th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>OncogenicityLastEvaluated</th>\n",
       "      <th>ReviewStatusOncogenicity</th>\n",
       "      <th>SCVsForAggregateGermlineClassification</th>\n",
       "      <th>SCVsForAggregateSomaticClinicalImpact</th>\n",
       "      <th>SCVsForAggregateOncogenicityClassification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 17, 2024</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV001451119|SCV005622007|SCV005909190</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Dec 17, 2024</td>\n",
       "      <td>397704705</td>\n",
       "      <td>...</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV001451119|SCV005622007|SCV005909190</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV000020156</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV000020156</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Jun 29, 2015</td>\n",
       "      <td>150829393</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV000020157</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0      15041                      Indel   \n",
       "1      15041                      Indel   \n",
       "2      15042                   Deletion   \n",
       "3      15042                   Deletion   \n",
       "4      15043  single nucleotide variant   \n",
       "\n",
       "                                                Name  GeneID GeneSymbol  \\\n",
       "0  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "1  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...    9907      AP5Z1   \n",
       "2     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "3     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1   \n",
       "4       NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "\n",
       "      HGNC_ID          ClinicalSignificance  ClinSigSimple LastEvaluated  \\\n",
       "0  HGNC:22197  Pathogenic/Likely pathogenic              1  Dec 17, 2024   \n",
       "1  HGNC:22197  Pathogenic/Likely pathogenic              1  Dec 17, 2024   \n",
       "2  HGNC:22197                    Pathogenic              1  Jun 29, 2010   \n",
       "3  HGNC:22197                    Pathogenic              1  Jun 29, 2010   \n",
       "4  HGNC:28986        Uncertain significance              0  Jun 29, 2015   \n",
       "\n",
       "   RS# (dbSNP)  ...      AlternateAlleleVCF SomaticClinicalImpact  \\\n",
       "0    397704705  ...  TGCTGTAAACTGTAACTGTAAA                     -   \n",
       "1    397704705  ...  TGCTGTAAACTGTAACTGTAAA                     -   \n",
       "2    397704709  ...                       G                     -   \n",
       "3    397704709  ...                       G                     -   \n",
       "4    150829393  ...                       A                     -   \n",
       "\n",
       "  SomaticClinicalImpactLastEvaluated ReviewStatusClinicalImpact Oncogenicity  \\\n",
       "0                                  -                          -            -   \n",
       "1                                  -                          -            -   \n",
       "2                                  -                          -            -   \n",
       "3                                  -                          -            -   \n",
       "4                                  -                          -            -   \n",
       "\n",
       "  OncogenicityLastEvaluated ReviewStatusOncogenicity  \\\n",
       "0                         -                        -   \n",
       "1                         -                        -   \n",
       "2                         -                        -   \n",
       "3                         -                        -   \n",
       "4                         -                        -   \n",
       "\n",
       "   SCVsForAggregateGermlineClassification  \\\n",
       "0  SCV001451119|SCV005622007|SCV005909190   \n",
       "1  SCV001451119|SCV005622007|SCV005909190   \n",
       "2                            SCV000020156   \n",
       "3                            SCV000020156   \n",
       "4                            SCV000020157   \n",
       "\n",
       "  SCVsForAggregateSomaticClinicalImpact  \\\n",
       "0                                     -   \n",
       "1                                     -   \n",
       "2                                     -   \n",
       "3                                     -   \n",
       "4                                     -   \n",
       "\n",
       "   SCVsForAggregateOncogenicityClassification  \n",
       "0                                           -  \n",
       "1                                           -  \n",
       "2                                           -  \n",
       "3                                           -  \n",
       "4                                           -  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz\", \n",
    "                 compression='gzip', \n",
    "                 sep='\\t', \n",
    "                 low_memory=False)\n",
    "\n",
    "print(\"الشكل العام للبيانات:\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c57703c9-d715-434f-92e0-705f6731af98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 7452145 entries, 0 to 7452144\n",
      "Data columns (total 43 columns):\n",
      " #   Column                                      Dtype \n",
      "---  ------                                      ----- \n",
      " 0   #AlleleID                                   int64 \n",
      " 1   Type                                        object\n",
      " 2   Name                                        object\n",
      " 3   GeneID                                      int64 \n",
      " 4   GeneSymbol                                  object\n",
      " 5   HGNC_ID                                     object\n",
      " 6   ClinicalSignificance                        object\n",
      " 7   ClinSigSimple                               int64 \n",
      " 8   LastEvaluated                               object\n",
      " 9   RS# (dbSNP)                                 int64 \n",
      " 10  nsv/esv (dbVar)                             object\n",
      " 11  RCVaccession                                object\n",
      " 12  PhenotypeIDS                                object\n",
      " 13  PhenotypeList                               object\n",
      " 14  Origin                                      object\n",
      " 15  OriginSimple                                object\n",
      " 16  Assembly                                    object\n",
      " 17  ChromosomeAccession                         object\n",
      " 18  Chromosome                                  object\n",
      " 19  Start                                       int64 \n",
      " 20  Stop                                        int64 \n",
      " 21  ReferenceAllele                             object\n",
      " 22  AlternateAllele                             object\n",
      " 23  Cytogenetic                                 object\n",
      " 24  ReviewStatus                                object\n",
      " 25  NumberSubmitters                            int64 \n",
      " 26  Guidelines                                  object\n",
      " 27  TestedInGTR                                 object\n",
      " 28  OtherIDs                                    object\n",
      " 29  SubmitterCategories                         int64 \n",
      " 30  VariationID                                 int64 \n",
      " 31  PositionVCF                                 int64 \n",
      " 32  ReferenceAlleleVCF                          object\n",
      " 33  AlternateAlleleVCF                          object\n",
      " 34  SomaticClinicalImpact                       object\n",
      " 35  SomaticClinicalImpactLastEvaluated          object\n",
      " 36  ReviewStatusClinicalImpact                  object\n",
      " 37  Oncogenicity                                object\n",
      " 38  OncogenicityLastEvaluated                   object\n",
      " 39  ReviewStatusOncogenicity                    object\n",
      " 40  SCVsForAggregateGermlineClassification      object\n",
      " 41  SCVsForAggregateSomaticClinicalImpact       object\n",
      " 42  SCVsForAggregateOncogenicityClassification  object\n",
      "dtypes: int64(10), object(33)\n",
      "memory usage: 2.4+ GB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ccce95b9-5d4f-4041-a7ea-20fbd12c3db0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 500,000 rows -> kept 142,346 benign/LB rows so far...\n",
      "Processed 1,000,000 rows -> kept 275,037 benign/LB rows so far...\n",
      "Processed 1,500,000 rows -> kept 532,384 benign/LB rows so far...\n",
      "Processed 2,000,000 rows -> kept 740,324 benign/LB rows so far...\n",
      "Processed 2,500,000 rows -> kept 980,874 benign/LB rows so far...\n",
      "Processed 3,000,000 rows -> kept 1,288,142 benign/LB rows so far...\n",
      "Processed 3,500,000 rows -> kept 1,471,697 benign/LB rows so far...\n",
      "Processed 4,000,000 rows -> kept 1,652,686 benign/LB rows so far...\n",
      "Processed 4,500,000 rows -> kept 1,683,299 benign/LB rows so far...\n",
      "Processed 5,000,000 rows -> kept 1,867,054 benign/LB rows so far...\n",
      "Processed 5,500,000 rows -> kept 2,206,895 benign/LB rows so far...\n",
      "Processed 6,000,000 rows -> kept 2,280,911 benign/LB rows so far...\n",
      "Processed 6,500,000 rows -> kept 2,332,934 benign/LB rows so far...\n",
      "Processed 7,000,000 rows -> kept 2,505,172 benign/LB rows so far...\n",
      "Processed 7,448,531 rows -> kept 2,545,075 benign/LB rows so far...\n",
      "\n",
      "DONE. Wrote 2,545,075 rows to clinvar_benign_likebenign_20250809.tsv\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\"\"\"\n",
    "Extract Benign / Likely benign variants from ClinVar's variant_summary.txt.gz\n",
    "- Keeps ONLY: Oncogenicity, Name, Type, OriginSimple, Chromosome, ReferenceAllele, AlternateAllele\n",
    "- Chunked processing to save RAM\n",
    "\"\"\"\n",
    "\n",
    "import re\n",
    "from datetime import datetime\n",
    "from typing import Optional\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "URL = \"https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz\"\n",
    "\n",
    "\n",
    "OUT_TSV = f\"clinvar_benign_likebenign_{datetime.now():%Y%m%d}.tsv\"  \n",
    "\n",
    "\n",
    "CHUNK = 500_000  \n",
    "\n",
    "\n",
    "USECOLS = [\n",
    "    \"Name\", \"Type\", \"OriginSimple\", \"Chromosome\",\n",
    "    \"ReferenceAllele\", \"AlternateAllele\", \"ClinicalSignificance\",\n",
    "]\n",
    "\n",
    "def map_oncogenicity(clinsig: str) -> Optional[str]:\n",
    "    \"\"\"\n",
    "    يحوّل ClinicalSignificance النصي إلى Oncogenicity مستهدفة:\n",
    "    - 'Benign' أو 'Likely benign' فقط\n",
    "    - يستبعد أي سجلات فيها كلمات غير مرغوبة (pathogenic/VUS/conflicting/..)\n",
    "    \"\"\"\n",
    "    if not isinstance(clinsig, str):\n",
    "        return None\n",
    "\n",
    "    text = clinsig.strip().lower()\n",
    "    \n",
    "    text = re.sub(r'[\\/,;|]+', ' ', text)\n",
    "    text = re.sub(r'[_\\s]+', ' ', text).strip()\n",
    "\n",
    "    \n",
    "    bad_terms = (\n",
    "        \"pathogenic\", \"likely pathogenic\", \"uncertain significance\",\n",
    "        \"conflicting\", \"risk factor\", \"drug response\", \"protective\",\n",
    "        \"affects\", \"association\", \"not provided\"\n",
    "    )\n",
    "    if any(b in text for b in bad_terms):\n",
    "        return None\n",
    "\n",
    "    has_likely = (\"likely benign\" in text) or (\"likely_benign\" in text)\n",
    "    has_benign = re.search(r'(^|[^a-z])benign([^a-z]|$)', text) is not None\n",
    "\n",
    "    \n",
    "    if has_benign and has_likely:\n",
    "        return \"Benign\"\n",
    "    if has_likely:\n",
    "        return \"Likely benign\"\n",
    "    if has_benign:\n",
    "        return \"Benign\"\n",
    "\n",
    "    return None\n",
    "\n",
    "def process():\n",
    "    total_in = total_out = 0\n",
    "    first_write = True\n",
    "\n",
    "    reader = pd.read_csv(\n",
    "        URL,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "        usecols=USECOLS,\n",
    "        chunksize=CHUNK,\n",
    "        dtype=str,\n",
    "        low_memory=False\n",
    "    )\n",
    "\n",
    "    for chunk in reader:\n",
    "        total_in += len(chunk)\n",
    "\n",
    "        \n",
    "        chunk[\"ClinicalSignificance\"] = chunk[\"ClinicalSignificance\"].fillna(\"\").str.strip()\n",
    "\n",
    "        \n",
    "        chunk[\"Oncogenicity\"] = chunk[\"ClinicalSignificance\"].apply(map_oncogenicity)\n",
    "\n",
    "       \n",
    "        df = chunk[chunk[\"Oncogenicity\"].notna()].copy()\n",
    "\n",
    "        \n",
    "        df_out = df[[\n",
    "            \"Oncogenicity\", \"Name\", \"Type\", \"OriginSimple\",\n",
    "            \"Chromosome\", \"ReferenceAllele\", \"AlternateAllele\"\n",
    "        ]]\n",
    "\n",
    "        if len(df_out):\n",
    "            df_out.to_csv(\n",
    "                OUT_TSV,\n",
    "                sep=\"\\t\",\n",
    "                index=False,\n",
    "                mode=\"w\" if first_write else \"a\",\n",
    "                header=first_write\n",
    "            )\n",
    "            first_write = False\n",
    "            total_out += len(df_out)\n",
    "\n",
    "        print(f\"Processed {total_in:,} rows -> kept {total_out:,} benign/LB rows so far...\")\n",
    "\n",
    "    print(f\"\\nDONE. Wrote {total_out:,} rows to {OUT_TSV}\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    process()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0f522398-7a3f-4566-895a-964bc680bc47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Benign    2545075\n",
       "Name: Oncogenicity, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df_kn = pd.read_csv(\"clinvar_benign_likebenign_20250809.tsv\", sep=\"\\t\", dtype=str)\n",
    "df_kn.head()\n",
    "df_kn[\"Oncogenicity\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "faae164c-fe5d-44b4-93b8-7f3016aab0d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_000410.4(HFE):c.892+48G&gt;A</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>6</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_000410.4(HFE):c.892+48G&gt;A</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>6</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NC_000007.14:g.128775556C&gt;T</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NC_000007.14:g.128775556C&gt;T</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_000355.4(TCN2):c.776G&gt;C (p.Arg259Pro)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>22</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2545070</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_015338.6(ASXL1):c.4469T&gt;C (p.Leu1490Ser)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>20</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2545071</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_004006.3(DMD):c.3202G&gt;A (p.Asp1068Asn)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>unknown</td>\n",
       "      <td>X</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2545072</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_004006.3(DMD):c.3202G&gt;A (p.Asp1068Asn)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>unknown</td>\n",
       "      <td>X</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2545073</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_001257291.2(SLC9A7):c.935C&gt;A (p.Thr312Asn)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>X</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2545074</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_001257291.2(SLC9A7):c.935C&gt;A (p.Thr312Asn)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>X</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2545075 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Oncogenicity                                           Name  \\\n",
       "0             Benign                   NM_000410.4(HFE):c.892+48G>A   \n",
       "1             Benign                   NM_000410.4(HFE):c.892+48G>A   \n",
       "2             Benign                    NC_000007.14:g.128775556C>T   \n",
       "3             Benign                    NC_000007.14:g.128775556C>T   \n",
       "4             Benign       NM_000355.4(TCN2):c.776G>C (p.Arg259Pro)   \n",
       "...              ...                                            ...   \n",
       "2545070       Benign    NM_015338.6(ASXL1):c.4469T>C (p.Leu1490Ser)   \n",
       "2545071       Benign      NM_004006.3(DMD):c.3202G>A (p.Asp1068Asn)   \n",
       "2545072       Benign      NM_004006.3(DMD):c.3202G>A (p.Asp1068Asn)   \n",
       "2545073       Benign  NM_001257291.2(SLC9A7):c.935C>A (p.Thr312Asn)   \n",
       "2545074       Benign  NM_001257291.2(SLC9A7):c.935C>A (p.Thr312Asn)   \n",
       "\n",
       "                              Type OriginSimple Chromosome ReferenceAllele  \\\n",
       "0        single nucleotide variant     germline          6              na   \n",
       "1        single nucleotide variant     germline          6              na   \n",
       "2        single nucleotide variant     germline          7              na   \n",
       "3        single nucleotide variant     germline          7              na   \n",
       "4        single nucleotide variant     germline         22              na   \n",
       "...                            ...          ...        ...             ...   \n",
       "2545070  single nucleotide variant     germline         20              na   \n",
       "2545071  single nucleotide variant      unknown          X              na   \n",
       "2545072  single nucleotide variant      unknown          X              na   \n",
       "2545073  single nucleotide variant     germline          X              na   \n",
       "2545074  single nucleotide variant     germline          X              na   \n",
       "\n",
       "        AlternateAllele  \n",
       "0                    na  \n",
       "1                    na  \n",
       "2                    na  \n",
       "3                    na  \n",
       "4                    na  \n",
       "...                 ...  \n",
       "2545070              na  \n",
       "2545071              na  \n",
       "2545072              na  \n",
       "2545073              na  \n",
       "2545074              na  \n",
       "\n",
       "[2545075 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df_kn = pd.read_csv(\"clinvar_benign_likebenign_20250809.tsv\", sep=\"\\t\", dtype=str)\n",
    "df_kn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cdf49a07-1639-4f7c-8cbe-ca4930b00bdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2545075 entries, 0 to 2545074\n",
      "Data columns (total 7 columns):\n",
      " #   Column           Dtype \n",
      "---  ------           ----- \n",
      " 0   Oncogenicity     object\n",
      " 1   Name             object\n",
      " 2   Type             object\n",
      " 3   OriginSimple     object\n",
      " 4   Chromosome       object\n",
      " 5   ReferenceAllele  object\n",
      " 6   AlternateAllele  object\n",
      "dtypes: object(7)\n",
      "memory usage: 135.9+ MB\n"
     ]
    }
   ],
   "source": [
    "df_kn.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e8b6ed4f-480c-47e3-b5b0-2d3378bf18f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Oncogenic    663330\n",
      "Name: Oncogenicity, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df_saratani['Oncogenicity'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "31febb1d-70ba-48c7-ad5b-11bd2bc8c7fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "na       2545059\n",
      "CTG            2\n",
      "AT             2\n",
      "G              2\n",
      "C              2\n",
      "CAG            2\n",
      "-              1\n",
      "A              1\n",
      "TTG            1\n",
      "AAAAT          1\n",
      "AAAA           1\n",
      "T              1\n",
      "Name: ReferenceAllele, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df_kn['ReferenceAllele'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5db6540c-0304-4cd6-9566-f6338f5eb28c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 500,000 rows -> kept 111,807 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 1,000,000 rows -> kept 211,369 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 1,500,000 rows -> kept 252,494 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 2,000,000 rows -> kept 316,718 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 2,500,000 rows -> kept 357,866 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 3,000,000 rows -> kept 388,277 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 3,500,000 rows -> kept 433,578 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 4,000,000 rows -> kept 461,955 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 4,500,000 rows -> kept 474,724 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 5,000,000 rows -> kept 530,529 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 5,500,000 rows -> kept 564,313 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 6,000,000 rows -> kept 589,112 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 6,500,000 rows -> kept 610,877 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 7,000,000 rows -> kept 652,962 oncogenic/Likely oncogenic rows so far...\n",
      "Processed 7,448,531 rows -> kept 663,330 oncogenic/Likely oncogenic rows so far...\n",
      "\n",
      "DONE. Wrote 663,330 rows to clinvar_oncogenic_likely_20250810.tsv\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\"\"\"\n",
    "Extract Oncogenic / Likely oncogenic variants from ClinVar's variant_summary.txt.gz\n",
    "Outputs columns aligned with your training:\n",
    "Oncogenicity, Name, Type, OriginSimple, Chromosome, ReferenceAllele, AlternateAllele\n",
    "\"\"\"\n",
    "\n",
    "import re\n",
    "from datetime import datetime\n",
    "from typing import Optional\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "URL = \"https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/variant_summary.txt.gz\"\n",
    "\n",
    "\n",
    "OUT_TSV = f\"clinvar_oncogenic_likely_{datetime.now():%Y%m%d}.tsv\"\n",
    "\n",
    "\n",
    "CHUNK = 500_000 \n",
    "\n",
    "\n",
    "USECOLS = [\n",
    "    \"Name\", \"Type\", \"OriginSimple\", \"Chromosome\",\n",
    "    \"ReferenceAllele\", \"AlternateAllele\", \"ClinicalSignificance\",\n",
    "]\n",
    "\n",
    "def map_oncogenicity_pos(clinsig: str) -> Optional[str]:\n",
    "    \"\"\"\n",
    "    Map ClinVar ClinicalSignificance to {Oncogenic, Likely oncogenic}\n",
    "    Excludes benign/uncertain/conflicting/risk/drug/protective/association/affects/not provided.\n",
    "    \"\"\"\n",
    "    if not isinstance(clinsig, str):\n",
    "        return None\n",
    "\n",
    "    text = clinsig.strip().lower()\n",
    "  \n",
    "    text = re.sub(r'[\\/,;|]+', ' ', text)\n",
    "    text = re.sub(r'[_\\s]+', ' ', text).strip()\n",
    "\n",
    "    \n",
    "    bad_terms = (\n",
    "        \"uncertain significance\", \"conflicting\", \"not provided\",\n",
    "        \"risk factor\", \"drug response\", \"protective\",\n",
    "        \"association\", \"affects\"\n",
    "    )\n",
    "    if any(b in text for b in bad_terms):\n",
    "        return None\n",
    "  \n",
    "    if \"benign\" in text:\n",
    "        return None\n",
    "\n",
    "    has_likely = (\"likely pathogenic\" in text) or (\"likely_pathogenic\" in text)\n",
    "   \n",
    "    has_path = re.search(r'(^|[^a-z])pathogenic([^a-z]|$)', text) is not None\n",
    "\n",
    "    if has_path and has_likely:\n",
    "        return \"Oncogenic\"\n",
    "    if has_path:\n",
    "        return \"Oncogenic\"\n",
    "    if has_likely:\n",
    "        return \"Likely oncogenic\"\n",
    "    return None\n",
    "\n",
    "def process():\n",
    "    total_in = total_out = 0\n",
    "    first_write = True\n",
    "\n",
    "    reader = pd.read_csv(\n",
    "        URL,\n",
    "        sep=\"\\t\",\n",
    "        compression=\"gzip\",\n",
    "        usecols=USECOLS,\n",
    "        chunksize=CHUNK,\n",
    "        dtype=str,\n",
    "        low_memory=False\n",
    "    )\n",
    "\n",
    "    for chunk in reader:\n",
    "        total_in += len(chunk)\n",
    "\n",
    "        chunk[\"ClinicalSignificance\"] = chunk[\"ClinicalSignificance\"].fillna(\"\").str.strip()\n",
    "        chunk[\"Oncogenicity\"] = chunk[\"ClinicalSignificance\"].apply(map_oncogenicity_pos)\n",
    "\n",
    "        df = chunk[chunk[\"Oncogenicity\"].notna()].copy()\n",
    "\n",
    "        df_out = df[[\n",
    "            \"Oncogenicity\", \"Name\", \"Type\", \"OriginSimple\",\n",
    "            \"Chromosome\", \"ReferenceAllele\", \"AlternateAllele\"\n",
    "        ]]\n",
    "\n",
    "        if len(df_out):\n",
    "            df_out.to_csv(\n",
    "                OUT_TSV,\n",
    "                sep=\"\\t\",\n",
    "                index=False,\n",
    "                mode=\"w\" if first_write else \"a\",\n",
    "                header=first_write\n",
    "            )\n",
    "            first_write = False\n",
    "            total_out += len(df_out)\n",
    "\n",
    "        print(f\"Processed {total_in:,} rows -> kept {total_out:,} oncogenic/Likely oncogenic rows so far...\")\n",
    "\n",
    "    print(f\"\\nDONE. Wrote {total_out:,} rows to {OUT_TSV}\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    process()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "74037f90-c909-4564-830d-f3f0a3eb4184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>Indel</td>\n",
       "      <td>germline</td>\n",
       "      <td>7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...</td>\n",
       "      <td>Indel</td>\n",
       "      <td>germline</td>\n",
       "      <td>7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>germline</td>\n",
       "      <td>7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>germline</td>\n",
       "      <td>7</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>11</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663325</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_002181.4(IHH):c.1018G&gt;A (p.Val340Met)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663326</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_004239.4(TRIP11):c.5272C&gt;T (p.Gln1758Ter)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>14</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663327</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_004239.4(TRIP11):c.5272C&gt;T (p.Gln1758Ter)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>14</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663328</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_000376.3(VDR):c.379G&gt;T (p.Glu127Ter)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>12</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663329</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_000376.3(VDR):c.379G&gt;T (p.Glu127Ter)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>12</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>663330 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Oncogenicity                                               Name  \\\n",
       "0         Oncogenic  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...   \n",
       "1         Oncogenic  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTA...   \n",
       "2         Oncogenic     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)   \n",
       "3         Oncogenic     NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)   \n",
       "4         Oncogenic        NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   \n",
       "...             ...                                                ...   \n",
       "663325    Oncogenic           NM_002181.4(IHH):c.1018G>A (p.Val340Met)   \n",
       "663326    Oncogenic       NM_004239.4(TRIP11):c.5272C>T (p.Gln1758Ter)   \n",
       "663327    Oncogenic       NM_004239.4(TRIP11):c.5272C>T (p.Gln1758Ter)   \n",
       "663328    Oncogenic            NM_000376.3(VDR):c.379G>T (p.Glu127Ter)   \n",
       "663329    Oncogenic            NM_000376.3(VDR):c.379G>T (p.Glu127Ter)   \n",
       "\n",
       "                             Type OriginSimple Chromosome ReferenceAllele  \\\n",
       "0                           Indel     germline          7              na   \n",
       "1                           Indel     germline          7              na   \n",
       "2                        Deletion     germline          7              na   \n",
       "3                        Deletion     germline          7              na   \n",
       "4       single nucleotide variant     germline         11              na   \n",
       "...                           ...          ...        ...             ...   \n",
       "663325  single nucleotide variant     germline          2              na   \n",
       "663326  single nucleotide variant     germline         14              na   \n",
       "663327  single nucleotide variant     germline         14              na   \n",
       "663328  single nucleotide variant     germline         12              na   \n",
       "663329  single nucleotide variant     germline         12              na   \n",
       "\n",
       "       AlternateAllele  \n",
       "0                   na  \n",
       "1                   na  \n",
       "2                   na  \n",
       "3                   na  \n",
       "4                   na  \n",
       "...                ...  \n",
       "663325              na  \n",
       "663326              na  \n",
       "663327              na  \n",
       "663328              na  \n",
       "663329              na  \n",
       "\n",
       "[663330 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df_saratani = pd.read_csv(\"clinvar_oncogenic_likely_20250810.tsv\", sep=\"\\t\", dtype=str)\n",
    "df_saratani"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4915ffce-6709-4beb-baf1-ba0d821c3872",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 663330 entries, 0 to 663329\n",
      "Data columns (total 7 columns):\n",
      " #   Column           Non-Null Count   Dtype \n",
      "---  ------           --------------   ----- \n",
      " 0   Oncogenicity     663330 non-null  object\n",
      " 1   Name             663330 non-null  object\n",
      " 2   Type             663330 non-null  object\n",
      " 3   OriginSimple     663330 non-null  object\n",
      " 4   Chromosome       663330 non-null  object\n",
      " 5   ReferenceAllele  663330 non-null  object\n",
      " 6   AlternateAllele  663330 non-null  object\n",
      "dtypes: object(7)\n",
      "memory usage: 35.4+ MB\n"
     ]
    }
   ],
   "source": [
    "df_saratani.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f2ce02d8-8630-4ccd-93fb-eaa8a7f97584",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Oncogenic    663330\n",
      "Name: Oncogenicity, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df_saratani['Oncogenicity'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c3e868b0-6752-45d7-8657-f1d244113f84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "na                             663246\n",
      "-                                  65\n",
      "CGCGGGGCGGGG                        2\n",
      "T                                   2\n",
      "ATTCT                               2\n",
      "GGCCTG                              2\n",
      "CAG                                 2\n",
      "C                                   2\n",
      "GGC                                 2\n",
      "GGAAAGCATCTCTGGCTCACCATGTAA         1\n",
      "GAGTTACAATTTCGATG                   1\n",
      "CC                                  1\n",
      "G                                   1\n",
      "A                                   1\n",
      "Name: ReferenceAllele, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df_saratani['ReferenceAllele'].value_counts())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d2d536b4-b7e4-44ac-96b0-61697122df5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1200000, 8) 0    600000\n",
      "1    600000\n",
      "Name: y, dtype: int64\n",
      "✅ Wrote balanced sample to clinvar_balanced_600k_600k.tsv\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "BENIGN_TSV = \"clinvar_benign_likebenign_20250809.tsv\"\n",
    "ONCO_TSV   = \"clinvar_oncogenic_likely_20250810.tsv\"\n",
    "\n",
    "SAMPLE_SIZE = 600_000\n",
    "RANDOM_SEED = 42\n",
    "\n",
    "# --- load benign ---\n",
    "ben = pd.read_csv(BENIGN_TSV, sep=\"\\t\", low_memory=False)\n",
    "if len(ben) > SAMPLE_SIZE:\n",
    "    ben = ben.sample(SAMPLE_SIZE, random_state=RANDOM_SEED)\n",
    "ben[\"y\"] = 0\n",
    "\n",
    "# --- load oncogenic ---\n",
    "onc = pd.read_csv(ONCO_TSV, sep=\"\\t\", low_memory=False)\n",
    "if len(onc) > SAMPLE_SIZE:\n",
    "    onc = onc.sample(SAMPLE_SIZE, random_state=RANDOM_SEED)\n",
    "onc[\"y\"] = 1\n",
    "\n",
    "# --- merge + shuffle ---\n",
    "df = pd.concat([ben, onc], ignore_index=True)\n",
    "df = df.sample(frac=1.0, random_state=RANDOM_SEED).reset_index(drop=True)\n",
    "\n",
    "print(df.shape, df[\"y\"].value_counts())\n",
    "\n",
    "# --- save to file ---\n",
    "df.to_csv(\"clinvar_balanced_600k_600k.tsv\", sep=\"\\t\", index=False)\n",
    "print(\"✅ Wrote balanced sample to clinvar_balanced_600k_600k.tsv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e2e3f006-2579-490a-a697-c7349d4d3799",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ben' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[2], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mben\u001b[49m\u001b[38;5;241m.\u001b[39minfo()\n",
      "\u001b[1;31mNameError\u001b[0m: name 'ben' is not defined"
     ]
    }
   ],
   "source": [
    "ben.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7f1950c1-be47-4b42-aa61-d88bffc75243",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 600000 entries, 11813 to 24438\n",
      "Data columns (total 8 columns):\n",
      " #   Column           Non-Null Count   Dtype \n",
      "---  ------           --------------   ----- \n",
      " 0   Oncogenicity     600000 non-null  object\n",
      " 1   Name             600000 non-null  object\n",
      " 2   Type             600000 non-null  object\n",
      " 3   OriginSimple     600000 non-null  object\n",
      " 4   Chromosome       600000 non-null  object\n",
      " 5   ReferenceAllele  600000 non-null  object\n",
      " 6   AlternateAllele  600000 non-null  object\n",
      " 7   y                600000 non-null  int64 \n",
      "dtypes: int64(1), object(7)\n",
      "memory usage: 41.2+ MB\n"
     ]
    }
   ],
   "source": [
    "onc.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "984deceb-5c65-416a-a94d-3631d69b1299",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mdf\u001b[49m\u001b[38;5;241m.\u001b[39minfo()\n",
      "\u001b[1;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "63c3489a-42df-40b0-96a4-b832efba876d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_002645.4(PIK3C2A):c.1721T&gt;C (p.Val574Ala)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>11</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_004082.5(DCTN1):c.2754C&gt;T (p.Pro918=)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_017799.4(TMEM260):c.2036C&gt;T (p.Pro679Leu)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>14</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_000085.5(CLCNKB):c.1846-1G&gt;A</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_000188.3(HK1):c.375+15C&gt;T</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>10</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Oncogenicity                                          Name  \\\n",
       "0       Benign  NM_002645.4(PIK3C2A):c.1721T>C (p.Val574Ala)   \n",
       "1       Benign      NM_004082.5(DCTN1):c.2754C>T (p.Pro918=)   \n",
       "2       Benign  NM_017799.4(TMEM260):c.2036C>T (p.Pro679Leu)   \n",
       "3    Oncogenic               NM_000085.5(CLCNKB):c.1846-1G>A   \n",
       "4       Benign                  NM_000188.3(HK1):c.375+15C>T   \n",
       "\n",
       "                        Type OriginSimple Chromosome ReferenceAllele  \\\n",
       "0  single nucleotide variant     germline         11              na   \n",
       "1  single nucleotide variant     germline          2              na   \n",
       "2  single nucleotide variant     germline         14              na   \n",
       "3  single nucleotide variant     germline          1              na   \n",
       "4  single nucleotide variant     germline         10              na   \n",
       "\n",
       "  AlternateAllele  y  \n",
       "0              na  0  \n",
       "1              na  0  \n",
       "2              na  0  \n",
       "3              na  1  \n",
       "4              na  0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "75406eb0-5b2b-4d03-bf76-7f3e80343ff8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    600000\n",
      "1    600000\n",
      "Name: y, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['y'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "92a1f2ea-670c-4e62-8e1a-6ab2dfc85e4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "الشكل العام للبيانات:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_002645.4(PIK3C2A):c.1721T&gt;C (p.Val574Ala)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>11</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_004082.5(DCTN1):c.2754C&gt;T (p.Pro918=)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_017799.4(TMEM260):c.2036C&gt;T (p.Pro679Leu)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>14</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_000085.5(CLCNKB):c.1846-1G&gt;A</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_000188.3(HK1):c.375+15C&gt;T</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>10</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Oncogenicity                                          Name  \\\n",
       "0       Benign  NM_002645.4(PIK3C2A):c.1721T>C (p.Val574Ala)   \n",
       "1       Benign      NM_004082.5(DCTN1):c.2754C>T (p.Pro918=)   \n",
       "2       Benign  NM_017799.4(TMEM260):c.2036C>T (p.Pro679Leu)   \n",
       "3    Oncogenic               NM_000085.5(CLCNKB):c.1846-1G>A   \n",
       "4       Benign                  NM_000188.3(HK1):c.375+15C>T   \n",
       "\n",
       "                        Type OriginSimple Chromosome ReferenceAllele  \\\n",
       "0  single nucleotide variant     germline         11              na   \n",
       "1  single nucleotide variant     germline          2              na   \n",
       "2  single nucleotide variant     germline         14              na   \n",
       "3  single nucleotide variant     germline          1              na   \n",
       "4  single nucleotide variant     germline         10              na   \n",
       "\n",
       "  AlternateAllele  y  \n",
       "0              na  0  \n",
       "1              na  0  \n",
       "2              na  0  \n",
       "3              na  1  \n",
       "4              na  0  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"clinvar_balanced_600k_600k.tsv\", \n",
    "                  \n",
    "                 sep='\\t', \n",
    "                 low_memory=False)\n",
    "\n",
    "print(\"الشكل العام للبيانات:\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ce2f563e-fdcc-4ac4-b30f-4cbed50ec3ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def parse_hgvs(name):\n",
    "    \"\"\"\n",
    "    Extract ref and alt alleles from HGVS-like strings.\n",
    "    Works for SNVs, splice site notations, etc.\n",
    "    \"\"\"\n",
    "    if not isinstance(name, str):\n",
    "        return None, None\n",
    "    \n",
    "    # simple SNV pattern: c.XXXXREF>ALT\n",
    "    m = re.search(r\"c\\.[0-9+-]*([ACGT])>([ACGT])\", name)\n",
    "    if m:\n",
    "        return m.group(1), m.group(2)\n",
    "    \n",
    "    # fallback: if it's like c.1234delA / c.1234insG\n",
    "    m = re.search(r\"c\\.[0-9+-]*del([ACGT]+)\", name)\n",
    "    if m:\n",
    "        return m.group(1), \"-\"\n",
    "    m = re.search(r\"c\\.[0-9+-]*ins([ACGT]+)\", name)\n",
    "    if m:\n",
    "        return \"-\", m.group(1)\n",
    "    \n",
    "    return None, None\n",
    "\n",
    "df[[\"Ref_from_Name\", \"Alt_from_Name\"]] = df[\"Name\"].apply(lambda x: pd.Series(parse_hgvs(x)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "812cbb89-2f0f-4e0b-8880-b81cd5a1aff5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>y</th>\n",
       "      <th>Ref_from_Name</th>\n",
       "      <th>Alt_from_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_002645.4(PIK3C2A):c.1721T&gt;C (p.Val574Ala)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>11</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_004082.5(DCTN1):c.2754C&gt;T (p.Pro918=)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>2</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_017799.4(TMEM260):c.2036C&gt;T (p.Pro679Leu)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>14</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_000085.5(CLCNKB):c.1846-1G&gt;A</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>1</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_000188.3(HK1):c.375+15C&gt;T</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>10</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Oncogenicity                                          Name  \\\n",
       "0       Benign  NM_002645.4(PIK3C2A):c.1721T>C (p.Val574Ala)   \n",
       "1       Benign      NM_004082.5(DCTN1):c.2754C>T (p.Pro918=)   \n",
       "2       Benign  NM_017799.4(TMEM260):c.2036C>T (p.Pro679Leu)   \n",
       "3    Oncogenic               NM_000085.5(CLCNKB):c.1846-1G>A   \n",
       "4       Benign                  NM_000188.3(HK1):c.375+15C>T   \n",
       "\n",
       "                        Type OriginSimple Chromosome ReferenceAllele  \\\n",
       "0  single nucleotide variant     germline         11              na   \n",
       "1  single nucleotide variant     germline          2              na   \n",
       "2  single nucleotide variant     germline         14              na   \n",
       "3  single nucleotide variant     germline          1              na   \n",
       "4  single nucleotide variant     germline         10              na   \n",
       "\n",
       "  AlternateAllele  y Ref_from_Name Alt_from_Name  \n",
       "0              na  0             T             C  \n",
       "1              na  0             C             T  \n",
       "2              na  0             C             T  \n",
       "3              na  1             G             A  \n",
       "4              na  0             C             T  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dcafbec6-bfe1-4de4-ad84-40d7b1e231ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C                   303968\n",
      "G                   286318\n",
      "T                   141360\n",
      "A                   132075\n",
      "-                        8\n",
      "CT                       1\n",
      "ACCGCTCGGCCCCCAG         1\n",
      "TTCC                     1\n",
      "Name: Ref_from_Name, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['Ref_from_Name'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b5bdf25b-ea70-4765-8f8f-3e798806bd0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T     301843\n",
      "A     249540\n",
      "G     160349\n",
      "C     151764\n",
      "-        235\n",
      "TG         1\n",
      "Name: Alt_from_Name, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['Alt_from_Name'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5649cc76-8171-4031-be72-325ff809ac37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 864224 entries, 0 to 864223\n",
      "Data columns (total 16 columns):\n",
      " #   Column           Non-Null Count   Dtype  \n",
      "---  ------           --------------   -----  \n",
      " 0   Oncogenicity     864224 non-null  object \n",
      " 1   Name             864224 non-null  object \n",
      " 2   Type             864224 non-null  object \n",
      " 3   OriginSimple     864224 non-null  object \n",
      " 4   Chromosome       864224 non-null  object \n",
      " 5   ReferenceAllele  863816 non-null  object \n",
      " 6   AlternateAllele  864164 non-null  object \n",
      " 7   y                864224 non-null  int64  \n",
      " 8   ref_len          863816 non-null  float64\n",
      " 9   alt_len          864164 non-null  float64\n",
      " 10  is_del           864224 non-null  int8   \n",
      " 11  is_ins           864224 non-null  int8   \n",
      " 12  is_snv           864224 non-null  int8   \n",
      " 13  is_mnv           864224 non-null  int8   \n",
      " 14  is_transition    864224 non-null  int64  \n",
      " 15  indel_len        864224 non-null  int32  \n",
      "dtypes: float64(2), int32(1), int64(2), int8(4), object(7)\n",
      "memory usage: 79.1+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a6a095d2-6d62-4225-9792-bb2ca88f25de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "germline            1130303\n",
      "unknown               63832\n",
      "germline/somatic       2234\n",
      "somatic                2099\n",
      "not provided           1473\n",
      "not applicable           59\n",
      "Name: OriginSimple, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['OriginSimple'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2872179c-df7e-47a4-a0e4-d3023f2fb9f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Benign       551425\n",
      "Oncogenic    312799\n",
      "Name: Oncogenicity, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['Oncogenicity'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8302e158-661a-469a-8465-948e9af16202",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>y</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>alt_len</th>\n",
       "      <th>is_del</th>\n",
       "      <th>is_ins</th>\n",
       "      <th>is_snv</th>\n",
       "      <th>is_mnv</th>\n",
       "      <th>is_transition</th>\n",
       "      <th>indel_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_002645.4(PIK3C2A):c.1721T&gt;C (p.Val574Ala)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>11</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_004082.5(DCTN1):c.2754C&gt;T (p.Pro918=)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_017799.4(TMEM260):c.2036C&gt;T (p.Pro679Leu)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_000085.5(CLCNKB):c.1846-1G&gt;A</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_000188.3(HK1):c.375+15C&gt;T</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>10</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Oncogenicity                                          Name  \\\n",
       "0       Benign  NM_002645.4(PIK3C2A):c.1721T>C (p.Val574Ala)   \n",
       "1       Benign      NM_004082.5(DCTN1):c.2754C>T (p.Pro918=)   \n",
       "2       Benign  NM_017799.4(TMEM260):c.2036C>T (p.Pro679Leu)   \n",
       "3    Oncogenic               NM_000085.5(CLCNKB):c.1846-1G>A   \n",
       "4       Benign                  NM_000188.3(HK1):c.375+15C>T   \n",
       "\n",
       "                        Type OriginSimple Chromosome ReferenceAllele  \\\n",
       "0  single nucleotide variant     germline         11               T   \n",
       "1  single nucleotide variant     germline          2               C   \n",
       "2  single nucleotide variant     germline         14               C   \n",
       "3  single nucleotide variant     germline          1               G   \n",
       "4  single nucleotide variant     germline         10               C   \n",
       "\n",
       "  AlternateAllele  y  ref_len  alt_len  is_del  is_ins  is_snv  is_mnv  \\\n",
       "0               C  0      1.0      1.0       0       0       1       0   \n",
       "1               T  0      1.0      1.0       0       0       1       0   \n",
       "2               T  0      1.0      1.0       0       0       1       0   \n",
       "3               A  1      1.0      1.0       0       0       1       0   \n",
       "4               T  0      1.0      1.0       0       0       1       0   \n",
       "\n",
       "   is_transition  indel_len  \n",
       "0              1          0  \n",
       "1              1          0  \n",
       "2              1          0  \n",
       "3              1          0  \n",
       "4              1          0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "928d1d27-9648-4166-9242-7919554fc870",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.0     863798\n",
      "3.0          6\n",
      "2.0          3\n",
      "5.0          2\n",
      "12.0         2\n",
      "16.0         1\n",
      "27.0         1\n",
      "4.0          1\n",
      "6.0          1\n",
      "17.0         1\n",
      "Name: ref_len, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['ref_len'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "988f9c8a-ad3d-4f9f-a4d0-a9439b0a4308",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T                       301849\n",
      "A                       249545\n",
      "G                       160356\n",
      "C                       151767\n",
      "-                          612\n",
      "CA                           4\n",
      "AATTAAGGTATA                 2\n",
      "GG                           2\n",
      "TG                           2\n",
      "TCCCGGGTTCAAGCGATTCT         2\n",
      "TTTCCGACAAAGGT               2\n",
      "ATAAATCACTTAGAGATGT          2\n",
      "AC                           2\n",
      "GTGG                         2\n",
      "TGACATCAGTCCGGGCAC           2\n",
      "TC                           2\n",
      "CTA                          2\n",
      "CACAAAGTG                    2\n",
      "TTGAA                        2\n",
      "TGG                          1\n",
      "AGTTACC                      1\n",
      "ACC                          1\n",
      "CTTA                         1\n",
      "TCT                          1\n",
      "Name: AlternateAllele, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['AlternateAllele'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d0da9091-c839-4b8c-a143-6940c4784e81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C                              303972\n",
      "G                              286320\n",
      "T                              141362\n",
      "A                              132075\n",
      "-                                  69\n",
      "CAG                                 3\n",
      "ATTCT                               2\n",
      "CGCGGGGCGGGG                        2\n",
      "GGC                                 2\n",
      "GGAAAGCATCTCTGGCTCACCATGTAA         1\n",
      "CC                                  1\n",
      "GGCCTG                              1\n",
      "TTCC                                1\n",
      "CTG                                 1\n",
      "ACCGCTCGGCCCCCAG                    1\n",
      "CT                                  1\n",
      "AT                                  1\n",
      "GAGTTACAATTTCGATG                   1\n",
      "Name: ReferenceAllele, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['ReferenceAllele'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7b0bd9ca-66a5-4732-9a58-099084e75804",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C                              303972\n",
      "G                              286320\n",
      "T                              141362\n",
      "A                              132075\n",
      "-                                  69\n",
      "CAG                                 3\n",
      "ATTCT                               2\n",
      "CGCGGGGCGGGG                        2\n",
      "GGC                                 2\n",
      "GGAAAGCATCTCTGGCTCACCATGTAA         1\n",
      "CC                                  1\n",
      "GGCCTG                              1\n",
      "TTCC                                1\n",
      "CTG                                 1\n",
      "ACCGCTCGGCCCCCAG                    1\n",
      "CT                                  1\n",
      "AT                                  1\n",
      "GAGTTACAATTTCGATG                   1\n",
      "Name: ReferenceAllele, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['ReferenceAllele'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "388b3fad-9488-4f69-a22a-9bca30a235dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\tahas\\AppData\\Local\\Temp\\ipykernel_71768\\2479845415.py:123: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  df[\"is_splice_core\"] = df[\"Name\"].astype(str).str.contains(splice_core_re, regex=True, na=False).astype(\"int8\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropped rows (missing/same alleles): 0\n",
      "\n",
      "AlternateAllele top values:\n",
      "T                       301849\n",
      "A                       249546\n",
      "G                       160356\n",
      "C                       151767\n",
      "-                          612\n",
      "NaN                         59\n",
      "CA                           4\n",
      "CACAAAGTG                    2\n",
      "AC                           2\n",
      "GG                           2\n",
      "TG                           2\n",
      "TCCCGGGTTCAAGCGATTCT         2\n",
      "TTTCCGACAAAGGT               2\n",
      "CTA                          2\n",
      "ATAAATCACTTAGAGATGT          2\n",
      "AATTAAGGTATA                 2\n",
      "GTGG                         2\n",
      "TTGAA                        2\n",
      "TGACATCAGTCCGGGCAC           2\n",
      "TC                           2\n",
      "Name: AlternateAllele, dtype: int64\n",
      "\n",
      "ReferenceAllele top values:\n",
      "C                              303972\n",
      "G                              286320\n",
      "T                              141362\n",
      "A                              132075\n",
      "NaN                               408\n",
      "-                                  69\n",
      "CAG                                 3\n",
      "ATTCT                               2\n",
      "CGCGGGGCGGGG                        2\n",
      "GGC                                 2\n",
      "GGAAAGCATCTCTGGCTCACCATGTAA         1\n",
      "CC                                  1\n",
      "GGCCTG                              1\n",
      "TTCC                                1\n",
      "CT                                  1\n",
      "ACCGCTCGGCCCCCAG                    1\n",
      "CTG                                 1\n",
      "AT                                  1\n",
      "GAGTTACAATTTCGATG                   1\n",
      "Name: ReferenceAllele, dtype: int64\n",
      "\n",
      "Variant shape counts:\n",
      "         count\n",
      "SNV     863748\n",
      "MNV          1\n",
      "DEL        237\n",
      "INS         24\n",
      "DELINS       0\n",
      "\n",
      "Functional flags (from Name):\n",
      "                     count\n",
      "missense            297647\n",
      "synonymous          300438\n",
      "nonsense                 0\n",
      "frameshift              40\n",
      "splice_core(±1/±2)  192356\n"
     ]
    }
   ],
   "source": [
    "import re, numpy as np, pandas as pd\n",
    "\n",
    "# ---------------------------\n",
    "# 1) Normalize parsed alleles\n",
    "# ---------------------------\n",
    "def norm_token(x):\n",
    "    if x is None: return np.nan\n",
    "    s = str(x).strip()\n",
    "    if s.lower() in {\"\", \"na\", \"nan\", \"none\"}:\n",
    "        return np.nan\n",
    "    return s.upper()\n",
    "\n",
    "\n",
    "df[\"ReferenceAllele\"] = df[\"ReferenceAllele\"].map(norm_token)\n",
    "df[\"AlternateAllele\"] = df[\"AlternateAllele\"].map(norm_token)\n",
    "\n",
    "# ------------------------------------------------\n",
    "# 2) Robust HGVS parser (c. level) from Name col\n",
    "#    handles SNV, del, ins, delins (basic cases)\n",
    "# ------------------------------------------------\n",
    "snv_re     = re.compile(r\"c\\.[0-9]+[+-]?[0-9]*([ACGT])>([ACGT])\", re.I)\n",
    "del_re     = re.compile(r\"c\\.[0-9]+(?:_[0-9]+)?del([ACGT]+)?\", re.I)   # 'del' with optional ref bases\n",
    "ins_re     = re.compile(r\"c\\.[0-9]+(?:_[0-9]+)?ins([ACGT]+)\", re.I)\n",
    "delins_re  = re.compile(r\"c\\.[0-9]+(?:_[0-9]+)?delins([ACGT]+)\", re.I)\n",
    "\n",
    "def parse_hgvs_alleles(name: str):\n",
    "    if not isinstance(name, str): return (np.nan, np.nan, None)\n",
    "    m = snv_re.search(name)\n",
    "    if m:  # SNV\n",
    "        return (m.group(1).upper(), m.group(2).upper(), \"SNV\")\n",
    "    m = delins_re.search(name)\n",
    "    if m:\n",
    "        alt = m.group(1).upper()\n",
    "        # 'delins' means some unknown ref replaced by ALT; mark as DELINS (ref unknown length)\n",
    "        return (\"-\", alt, \"DELINS\")\n",
    "    m = del_re.search(name)\n",
    "    if m:\n",
    "        ref = m.group(1).upper() if m.group(1) else \"-\"  # sometimes ClinVar omits explicit ref\n",
    "        return (ref, \"-\", \"DEL\")\n",
    "    m = ins_re.search(name)\n",
    "    if m:\n",
    "        alt = m.group(1).upper()\n",
    "        return (\"-\", alt, \"INS\")\n",
    "    return (np.nan, np.nan, None)\n",
    "\n",
    "# Only fill where alleles are still missing\n",
    "need_ref = df[\"ReferenceAllele\"].isna()\n",
    "need_alt = df[\"AlternateAllele\"].isna()\n",
    "parsed = df.loc[need_ref | need_alt, \"Name\"].apply(parse_hgvs_alleles)\n",
    "if len(parsed):\n",
    "    pf = pd.DataFrame(list(parsed), index=parsed.index, columns=[\"_ref\",\"_alt\",\"_kind\"])\n",
    "    fill_ref = need_ref & pf[\"_ref\"].notna()\n",
    "    fill_alt = need_alt & pf[\"_alt\"].notna()\n",
    "    df.loc[fill_ref, \"ReferenceAllele\"] = pf.loc[fill_ref, \"_ref\"]\n",
    "    df.loc[fill_alt, \"AlternateAllele\"] = pf.loc[fill_alt, \"_alt\"]\n",
    "    df[\"_hgvs_kind\"] = pf[\"_kind\"]\n",
    "else:\n",
    "    df[\"_hgvs_kind\"] = None\n",
    "\n",
    "# ------------------------------------------------------\n",
    "# 3) Validate alleles are A/C/G/T strings or a single '-'\n",
    "# ------------------------------------------------------\n",
    "valid = set(\"ACGT\")\n",
    "def clean_allele(x):\n",
    "    if pd.isna(x): return np.nan\n",
    "    s = str(x).upper()\n",
    "    if s == \"-\": return s\n",
    "    if all(ch in valid for ch in s):\n",
    "        return s\n",
    "    return np.nan  # drop weird tokens\n",
    "\n",
    "df[\"ReferenceAllele\"] = df[\"ReferenceAllele\"].map(clean_allele)\n",
    "df[\"AlternateAllele\"] = df[\"AlternateAllele\"].map(clean_allele)\n",
    "\n",
    "# ----------------------------------------\n",
    "# 4) Consistency + derived variant features\n",
    "# ----------------------------------------\n",
    "df[\"ref_len\"] = df[\"ReferenceAllele\"].str.len()\n",
    "df[\"alt_len\"] = df[\"AlternateAllele\"].str.len()\n",
    "\n",
    "# invalid cases\n",
    "df[\"invalid_same\"] = (df[\"ReferenceAllele\"].notna() & (df[\"ReferenceAllele\"] == df[\"AlternateAllele\"]))\n",
    "df[\"invalid_both_missing\"] = df[\"ReferenceAllele\"].isna() & df[\"AlternateAllele\"].isna()\n",
    "\n",
    "# shape flags\n",
    "df[\"is_del\"] = ((df[\"AlternateAllele\"] == \"-\") & df[\"ReferenceAllele\"].notna()).astype(\"int8\")\n",
    "df[\"is_ins\"] = ((df[\"ReferenceAllele\"] == \"-\") & df[\"AlternateAllele\"].notna()).astype(\"int8\")\n",
    "df[\"is_snv\"] = ((df[\"ref_len\"] == 1) & (df[\"alt_len\"] == 1)).fillna(False).astype(\"int8\")\n",
    "df[\"is_mnv\"] = ((df[\"ref_len\"] > 1) & (df[\"alt_len\"] > 1)).fillna(False).astype(\"int8\")\n",
    "df[\"is_delins\"] = ((df[\"_hgvs_kind\"] == \"DELINS\") & df[\"AlternateAllele\"].notna()).astype(\"int8\")\n",
    "\n",
    "# transitions only for SNV\n",
    "transitions = {(\"A\",\"G\"),(\"G\",\"A\"),(\"C\",\"T\"),(\"T\",\"C\")}\n",
    "df[\"is_transition\"] = [\n",
    "    int(snv and ( (r,a) in transitions ))\n",
    "    for snv, r, a in zip(df[\"is_snv\"], df[\"ReferenceAllele\"], df[\"AlternateAllele\"])\n",
    "]\n",
    "\n",
    "# indel length (0 for SNV/MNV)\n",
    "df[\"indel_len\"] = np.where(\n",
    "    df[\"is_del\"].astype(bool) | df[\"is_ins\"].astype(bool),\n",
    "    (df[\"ref_len\"].fillna(0) - df[\"alt_len\"].fillna(0)).abs(),\n",
    "    0\n",
    ").astype(\"int32\")\n",
    "\n",
    "# -------------------------------------------\n",
    "# 5) Optional: functional flags from Name (p.)\n",
    "# -------------------------------------------\n",
    "# missense: p.Val574Ala, synonymous: p.Pro918=, nonsense: p.Tyr123*, frameshift: p.Lys10fs\n",
    "is_missense_re   = re.compile(r\"\\bp\\.[A-Z][a-z]{2}[0-9]+(?!TER|STOP|X)[A-Z][a-z]{2}\\b\", re.I)\n",
    "is_syn_re       = re.compile(r\"\\bp\\.[A-Z][a-z]{2}[0-9]+=\")\n",
    "is_nonsense_re   = re.compile(r\"\\bp\\.[A-Z][a-z]{2}[0-9]+(\\*|TER|STOP|X)\\b\", re.I)\n",
    "is_frameshift_re= re.compile(r\"\\bp\\.[A-Z][a-z]{2}[0-9]+fs\", re.I)\n",
    "\n",
    "# splice-proximal core (±1 or ±2)\n",
    "splice_core_re  = re.compile(r\"c\\.[0-9]+([+-][12])\", re.I)\n",
    "\n",
    "df[\"is_missense\"]   = df[\"Name\"].astype(str).str.contains(is_missense_re).astype(\"int8\")\n",
    "df[\"is_synonymous\"] = df[\"Name\"].astype(str).str.contains(is_syn_re).astype(\"int8\")\n",
    "df[\"is_nonsense\"]   = df[\"Name\"].astype(str).str.contains(is_nonsense_re).astype(\"int8\")\n",
    "df[\"is_frameshift\"] = df[\"Name\"].astype(str).str.contains(is_frameshift_re).astype(\"int8\")\n",
    "# fix the splice regex warning (and robust to NaNs)\n",
    "df[\"is_splice_core\"] = df[\"Name\"].astype(str).str.contains(splice_core_re, regex=True, na=False).astype(\"int8\")\n",
    "# compact SNV category (A>C etc.) for SNVs only\n",
    "df[\"snv_pair\"] = np.where(\n",
    "    (df[\"is_snv\"]==1) & df[\"ReferenceAllele\"].notna() & df[\"AlternateAllele\"].notna(),\n",
    "    df[\"ReferenceAllele\"].astype(str) + \">\" + df[\"AlternateAllele\"].astype(str),\n",
    "    \"NA\"\n",
    ")\n",
    "\n",
    "# --------------------------------\n",
    "# 6) Filter clearly invalid rows\n",
    "# --------------------------------\n",
    "before = len(df)\n",
    "df = df[~df[\"invalid_both_missing\"] & ~df[\"invalid_same\"]].reset_index(drop=True)\n",
    "after = len(df)\n",
    "\n",
    "# --------------------------------\n",
    "# 7) Diagnostics\n",
    "# --------------------------------\n",
    "print(\"Dropped rows (missing/same alleles):\", before - after)\n",
    "print(\"\\nAlternateAllele top values:\")\n",
    "print(df[\"AlternateAllele\"].value_counts(dropna=False).head(20))\n",
    "print(\"\\nReferenceAllele top values:\")\n",
    "print(df[\"ReferenceAllele\"].value_counts(dropna=False).head(20))\n",
    "\n",
    "print(\"\\nVariant shape counts:\")\n",
    "print(pd.DataFrame({\n",
    "    \"SNV\": df[\"is_snv\"].sum(),\n",
    "    \"MNV\": df[\"is_mnv\"].sum(),\n",
    "    \"DEL\": df[\"is_del\"].sum(),\n",
    "    \"INS\": df[\"is_ins\"].sum(),\n",
    "    \"DELINS\": df[\"is_delins\"].sum()\n",
    "}, index=[\"count\"]).T)\n",
    "\n",
    "print(\"\\nFunctional flags (from Name):\")\n",
    "print(pd.DataFrame({\n",
    "    \"missense\": df[\"is_missense\"].sum(),\n",
    "    \"synonymous\": df[\"is_synonymous\"].sum(),\n",
    "    \"nonsense\": df[\"is_nonsense\"].sum(),\n",
    "    \"frameshift\": df[\"is_frameshift\"].sum(),\n",
    "    \"splice_core(±1/±2)\": df[\"is_splice_core\"].sum(),\n",
    "}, index=[\"count\"]).T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "547be329-b804-42ce-ae03-aac9c608ce4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>y</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>alt_len</th>\n",
       "      <th>...</th>\n",
       "      <th>_hgvs_kind</th>\n",
       "      <th>invalid_same</th>\n",
       "      <th>invalid_both_missing</th>\n",
       "      <th>is_delins</th>\n",
       "      <th>is_missense</th>\n",
       "      <th>is_synonymous</th>\n",
       "      <th>is_nonsense</th>\n",
       "      <th>is_frameshift</th>\n",
       "      <th>is_splice_core</th>\n",
       "      <th>snv_pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_002645.4(PIK3C2A):c.1721T&gt;C (p.Val574Ala)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>11</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>T&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_004082.5(DCTN1):c.2754C&gt;T (p.Pro918=)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>2</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_017799.4(TMEM260):c.2036C&gt;T (p.Pro679Leu)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>14</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_000085.5(CLCNKB):c.1846-1G&gt;A</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_000188.3(HK1):c.375+15C&gt;T</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>10</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_001042492.3(NF1):c.4892T&gt;G (p.Leu1631Ter)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>17</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>T&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_021250.4(LILRA5):c.55G&gt;A (p.Val19Met)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>19</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_003985.6(TNK1):c.1299G&gt;A (p.Ser433=)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>17</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Benign</td>\n",
       "      <td>NM_001277115.2(DNAH11):c.11352C&gt;T (p.Phe3784=)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Oncogenic</td>\n",
       "      <td>NM_001267550.2(TTN):c.79162G&gt;T (p.Gly26388Ter)</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>germline</td>\n",
       "      <td>2</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>G&gt;T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Oncogenicity                                            Name  \\\n",
       "0       Benign    NM_002645.4(PIK3C2A):c.1721T>C (p.Val574Ala)   \n",
       "1       Benign        NM_004082.5(DCTN1):c.2754C>T (p.Pro918=)   \n",
       "2       Benign    NM_017799.4(TMEM260):c.2036C>T (p.Pro679Leu)   \n",
       "3    Oncogenic                 NM_000085.5(CLCNKB):c.1846-1G>A   \n",
       "4       Benign                    NM_000188.3(HK1):c.375+15C>T   \n",
       "5    Oncogenic    NM_001042492.3(NF1):c.4892T>G (p.Leu1631Ter)   \n",
       "6       Benign        NM_021250.4(LILRA5):c.55G>A (p.Val19Met)   \n",
       "7       Benign         NM_003985.6(TNK1):c.1299G>A (p.Ser433=)   \n",
       "8       Benign  NM_001277115.2(DNAH11):c.11352C>T (p.Phe3784=)   \n",
       "9    Oncogenic  NM_001267550.2(TTN):c.79162G>T (p.Gly26388Ter)   \n",
       "\n",
       "                        Type OriginSimple Chromosome ReferenceAllele  \\\n",
       "0  single nucleotide variant     germline         11               T   \n",
       "1  single nucleotide variant     germline          2               C   \n",
       "2  single nucleotide variant     germline         14               C   \n",
       "3  single nucleotide variant     germline          1               G   \n",
       "4  single nucleotide variant     germline         10               C   \n",
       "5  single nucleotide variant     germline         17               T   \n",
       "6  single nucleotide variant     germline         19               G   \n",
       "7  single nucleotide variant     germline         17               G   \n",
       "8  single nucleotide variant     germline          7               C   \n",
       "9  single nucleotide variant     germline          2               G   \n",
       "\n",
       "  AlternateAllele  y  ref_len  alt_len  ...  _hgvs_kind  invalid_same  \\\n",
       "0               C  0      1.0      1.0  ...         NaN         False   \n",
       "1               T  0      1.0      1.0  ...         NaN         False   \n",
       "2               T  0      1.0      1.0  ...         NaN         False   \n",
       "3               A  1      1.0      1.0  ...         NaN         False   \n",
       "4               T  0      1.0      1.0  ...         NaN         False   \n",
       "5               G  1      1.0      1.0  ...         NaN         False   \n",
       "6               A  0      1.0      1.0  ...         NaN         False   \n",
       "7               A  0      1.0      1.0  ...         NaN         False   \n",
       "8               T  0      1.0      1.0  ...         NaN         False   \n",
       "9               T  1      1.0      1.0  ...         NaN         False   \n",
       "\n",
       "   invalid_both_missing  is_delins  is_missense  is_synonymous is_nonsense  \\\n",
       "0                 False          0            1              0           0   \n",
       "1                 False          0            0              1           0   \n",
       "2                 False          0            1              0           0   \n",
       "3                 False          0            0              0           0   \n",
       "4                 False          0            0              0           0   \n",
       "5                 False          0            1              0           0   \n",
       "6                 False          0            1              0           0   \n",
       "7                 False          0            0              1           0   \n",
       "8                 False          0            0              1           0   \n",
       "9                 False          0            1              0           0   \n",
       "\n",
       "   is_frameshift  is_splice_core  snv_pair  \n",
       "0              0               0       T>C  \n",
       "1              0               0       C>T  \n",
       "2              0               0       C>T  \n",
       "3              0               1       G>A  \n",
       "4              0               1       C>T  \n",
       "5              0               0       T>G  \n",
       "6              0               0       G>A  \n",
       "7              0               0       G>A  \n",
       "8              0               0       C>T  \n",
       "9              0               0       G>T  \n",
       "\n",
       "[10 rows x 26 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d5f0576e-abf9-4f9e-8f5a-65cb89f195ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 864224 entries, 0 to 864223\n",
      "Data columns (total 26 columns):\n",
      " #   Column                Non-Null Count   Dtype  \n",
      "---  ------                --------------   -----  \n",
      " 0   Oncogenicity          864224 non-null  object \n",
      " 1   Name                  864224 non-null  object \n",
      " 2   Type                  864224 non-null  object \n",
      " 3   OriginSimple          864224 non-null  object \n",
      " 4   Chromosome            864224 non-null  object \n",
      " 5   ReferenceAllele       863816 non-null  object \n",
      " 6   AlternateAllele       864165 non-null  object \n",
      " 7   y                     864224 non-null  int64  \n",
      " 8   ref_len               863816 non-null  float64\n",
      " 9   alt_len               864165 non-null  float64\n",
      " 10  is_del                864224 non-null  int8   \n",
      " 11  is_ins                864224 non-null  int8   \n",
      " 12  is_snv                864224 non-null  int8   \n",
      " 13  is_mnv                864224 non-null  int8   \n",
      " 14  is_transition         864224 non-null  int64  \n",
      " 15  indel_len             864224 non-null  int32  \n",
      " 16  _hgvs_kind            1 non-null       object \n",
      " 17  invalid_same          864224 non-null  bool   \n",
      " 18  invalid_both_missing  864224 non-null  bool   \n",
      " 19  is_delins             864224 non-null  int8   \n",
      " 20  is_missense           864224 non-null  int8   \n",
      " 21  is_synonymous         864224 non-null  int8   \n",
      " 22  is_nonsense           864224 non-null  int8   \n",
      " 23  is_frameshift         864224 non-null  int8   \n",
      " 24  is_splice_core        864224 non-null  int8   \n",
      " 25  snv_pair              864224 non-null  object \n",
      "dtypes: bool(2), float64(2), int32(1), int64(2), int8(10), object(9)\n",
      "memory usage: 98.9+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e2feaabb-33eb-41da-9b60-4e9b2dcfa3d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Device: cuda\n",
      "Shapes: (864224, 4) (864224, 13) (864224,)\n"
     ]
    }
   ],
   "source": [
    "import os, json\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import roc_auc_score, average_precision_score\n",
    "import torch, torch.nn as nn\n",
    "from torch.utils.data import Dataset, DataLoader\n",
    "\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "print(\"Device:\", device)\n",
    "\n",
    "# features (no leakage)\n",
    "cat_cols = [\"Type\",\"OriginSimple\",\"Chromosome\",\"snv_pair\"]\n",
    "num_cols = [\"ref_len\",\"alt_len\",\"indel_len\",\"is_del\",\"is_ins\",\"is_snv\",\"is_mnv\",\n",
    "            \"is_transition\",\"is_missense\",\"is_synonymous\",\"is_nonsense\",\"is_frameshift\",\"is_splice_core\"]\n",
    "\n",
    "# tidy dtypes\n",
    "for c in cat_cols:\n",
    "    df[c] = df[c].astype(\"string\")\n",
    "for c in num_cols:\n",
    "    df[c] = pd.to_numeric(df[c], errors=\"coerce\")\n",
    "df[num_cols] = df[num_cols].fillna(0)\n",
    "\n",
    "y = df[\"y\"].astype(int).values\n",
    "\n",
    "# vocab + z-stats\n",
    "def build_vocab(series):\n",
    "    uniq = [\"UNK\"] + sorted({str(x) for x in series.fillna(\"UNK\")})\n",
    "    return {v:i for i,v in enumerate(uniq)}\n",
    "\n",
    "cat_maps = {c: build_vocab(df[c]) for c in cat_cols}\n",
    "num_stats = {c: (float(df[c].mean()), float(df[c].std() or 1.0)) for c in num_cols}\n",
    "\n",
    "# encoders\n",
    "def encode_cats(frame):\n",
    "    mats = []\n",
    "    for c in cat_cols:\n",
    "        vocab = cat_maps[c]\n",
    "        mats.append(frame[c].fillna(\"UNK\").astype(str).map(lambda v: vocab.get(v, 0)).astype(\"int64\").values)\n",
    "    return np.vstack(mats).T\n",
    "\n",
    "def zscore_nums(frame):\n",
    "    arrs = []\n",
    "    for c in num_cols:\n",
    "        m, s = num_stats[c]\n",
    "        arrs.append(((frame[c].values - m) / (s if s!=0 else 1.0)).astype(\"float32\"))\n",
    "    return np.vstack(arrs).T\n",
    "\n",
    "X_cat_all = encode_cats(df)\n",
    "X_num_all = zscore_nums(df)\n",
    "print(\"Shapes:\", X_cat_all.shape, X_num_all.shape, y.shape)\n",
    "\n",
    "# save artifacts for later deployment\n",
    "os.makedirs(\"mutonco/artifacts\", exist_ok=True)\n",
    "with open(\"mutonco/artifacts/cat_maps.json\",\"w\",encoding=\"utf-8\") as f: json.dump(cat_maps,f)\n",
    "with open(\"mutonco/artifacts/num_stats.json\",\"w\",encoding=\"utf-8\") as f: json.dump(num_stats,f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "5be21b76-edad-4433-afb2-d1d6a6c56b1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Wrote balanced sample to clinvar_final2.tsv\n"
     ]
    }
   ],
   "source": [
    "# --- save to file ---\n",
    "df.to_csv(\"clinvar_final2.tsv\", sep=\"\\t\", index=False)\n",
    "print(\"✅ Wrote balanced sample to clinvar_final2.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "213c75f9-fc14-48f5-8ddb-4e8e45e42be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns=[\"is_delins\",\"is_frameshift\",\"is_snv\",\"is_mnv\",\"ref_len\",\"alt_len\",\"indel_len\",\"invalid_same\",\"invalid_both_missing\",\"Oncogenicity\",\"is_del\",\"is_ins\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "e4907455-6533-4c3d-b380-9dc3af78a405",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop(columns=[\"is_del\",\"is_ins\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "33bfe2d4-e894-4b74-b947-2771121a9d55",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_excel('C:/Users/tahas/clinvar_final2.xlsx', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "62ad40e0-b827-4d84-b02b-e8ff6e60b330",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    557785\n",
      "1    279597\n",
      "Name: is_missense, dtype: int64\n",
      "0    837382\n",
      "Name: is_nonsense, dtype: int64\n",
      "0    650889\n",
      "1    186493\n",
      "Name: is_splice_core, dtype: int64\n",
      "C>T    211437\n",
      "G>A    177139\n",
      "T>C     85970\n",
      "A>G     84236\n",
      "G>T     58263\n",
      "C>G     44344\n",
      "C>A     41067\n",
      "G>C     40625\n",
      "T>G     27561\n",
      "T>A     23312\n",
      "A>T     22309\n",
      "A>C     21104\n",
      "->T         4\n",
      "->A         3\n",
      "->C         3\n",
      "->G         2\n",
      "NA          2\n",
      "C>-         1\n",
      "Name: snv_pair, dtype: Int64\n",
      "single nucleotide variant    837382\n",
      "Name: Type, dtype: Int64\n",
      "germline    837382\n",
      "Name: OriginSimple, dtype: Int64\n",
      "2     82640\n",
      "1     72571\n",
      "17    51384\n",
      "11    48985\n",
      "3     44758\n",
      "X     43819\n",
      "16    43414\n",
      "19    41260\n",
      "7     40225\n",
      "5     39434\n",
      "9     38186\n",
      "12    37705\n",
      "6     36233\n",
      "15    31734\n",
      "10    29219\n",
      "8     27555\n",
      "4     26623\n",
      "14    25244\n",
      "22    17356\n",
      "13    16909\n",
      "20    16800\n",
      "18    14720\n",
      "21    10272\n",
      "Y       320\n",
      "na       16\n",
      "Name: Chromosome, dtype: Int64\n",
      "C    296849\n",
      "G    276027\n",
      "T    136843\n",
      "A    127649\n",
      "-        14\n",
      "Name: ReferenceAllele, dtype: int64\n",
      "T       292013\n",
      "A       241521\n",
      "G       156143\n",
      "C       147702\n",
      "-            1\n",
      "ACC          1\n",
      "CTTA         1\n",
      "Name: AlternateAllele, dtype: int64\n",
      "1    558782\n",
      "0    278600\n",
      "Name: is_transition, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(df['is_missense'].value_counts())\n",
    "print(df['is_nonsense'].value_counts())\n",
    "print(df['is_splice_core'].value_counts())\n",
    "print(df['snv_pair'].value_counts())\n",
    "print(df['Type'].value_counts())\n",
    "print(df['OriginSimple'].value_counts())\n",
    "print(df['Chromosome'].value_counts())\n",
    "print(df['ReferenceAllele'].value_counts())\n",
    "print(df['AlternateAllele'].value_counts())\n",
    "print(df['is_transition'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "bc0aa5c3-9e16-4812-92e7-8fc069e8315b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 837382 entries, 0 to 837381\n",
      "Data columns (total 12 columns):\n",
      " #   Column           Non-Null Count   Dtype \n",
      "---  ------           --------------   ----- \n",
      " 0   Name             837382 non-null  object\n",
      " 1   Type             837382 non-null  string\n",
      " 2   OriginSimple     837382 non-null  string\n",
      " 3   Chromosome       837382 non-null  string\n",
      " 4   ReferenceAllele  837382 non-null  object\n",
      " 5   AlternateAllele  837382 non-null  object\n",
      " 6   y                837382 non-null  int64 \n",
      " 7   is_transition    837382 non-null  int64 \n",
      " 8   is_missense      837382 non-null  int8  \n",
      " 9   is_nonsense      837382 non-null  int8  \n",
      " 10  is_splice_core   837382 non-null  int8  \n",
      " 11  snv_pair         837382 non-null  string\n",
      "dtypes: int64(2), int8(3), object(3), string(4)\n",
      "memory usage: 59.9+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "5c91e97e-26c5-45a2-88e7-934909fe7912",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "germline    837382\n",
      "Name: OriginSimple, dtype: Int64\n",
      "single nucleotide variant    837382\n",
      "Name: Type, dtype: Int64\n"
     ]
    }
   ],
   "source": [
    "print(df['OriginSimple'].value_counts())\n",
    "print(df['Type'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "a7e48436-d427-4788-a8c6-f5554a8745e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kept 837,876 germline rows; dropped 26,348 others.\n",
      "germline    837876\n",
      "Name: OriginSimple, dtype: Int64\n"
     ]
    }
   ],
   "source": [
    "# احتفظ فقط بالسجلات التي OriginSimple فيها \"germline\" (مع تطبيع المسافات/الحروف)\n",
    "mask = df[\"OriginSimple\"].astype(str).str.strip().str.lower().eq(\"germline\")\n",
    "kept = int(mask.sum()); dropped = int(len(df) - kept)\n",
    "\n",
    "df = df.loc[mask].reset_index(drop=True)\n",
    "\n",
    "print(f\"Kept {kept:,} germline rows; dropped {dropped:,} others.\")\n",
    "print(df[\"OriginSimple\"].value_counts(dropna=False))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "cce83c4c-c90c-47f2-965e-1ca0bb3ceb29",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kept 837,382 germline rows; dropped 494 others.\n",
      "single nucleotide variant    837382\n",
      "Name: Type, dtype: Int64\n"
     ]
    }
   ],
   "source": [
    "# احتفظ فقط بالسجلات التي OriginSimple فيها \"germline\" (مع تطبيع المسافات/الحروف)\n",
    "mask = df[\"Type\"].astype(str).str.strip().str.lower().eq(\"single nucleotide variant\")\n",
    "kept = int(mask.sum()); dropped = int(len(df) - kept)\n",
    "\n",
    "df = df.loc[mask].reset_index(drop=True)\n",
    "\n",
    "print(f\"Kept {kept:,} germline rows; dropped {dropped:,} others.\")\n",
    "print(df[\"Type\"].value_counts(dropna=False))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
